ui/app/components/app/metamask-template-renderer/metamask-template-renderer.js

```diff
@@ -0,0 +1,64 @@
+import React, { memo } from 'react'
+import PropTypes from 'prop-types'
+import { safeComponentList } from './safe-component-list'
+
+const MetaMaskTemplateRenderer = ({ sections }) => {
+  if (!sections) {
+    return null
+  }
+  if (typeof sections === 'string') {
+    return sections
+  }
+  return (
+    <>
+      {sections.reduce((acc, child, index) => {
+        // React can render strings directly, so push them into the accumulator
+        if (typeof child === 'string') {
+          acc.push(child)
+        } else if (child) {
+          // The other option is one of our Sections, which contains
+          // element, children, and props.
+          const { element, children, props } = child
+          const Element = safeComponentList[element]
+          if (!Element) {
+            console.warn(
+              `${element} is not in the safe component list and will not be rendered`,
+            )
+          }
+          const childrenOrNull =
+            Element && children ? (
+              <MetaMaskTemplateRenderer sections={children} />
+            ) : null
+          if (Element) {
+            acc.push(
+              <Element key={`${element}${index + 0}`} {...props}>
+                {childrenOrNull}
+              </Element>,
+            )
+          }
+        }
+        return acc
+      }, [])}
+    </>
+  )
+}
+
+const SectionShape = {
+  props: PropTypes.object,
+  element: PropTypes.oneOf(Object.keys(safeComponentList)),
+}
+
+const ValidChildren = PropTypes.oneOfType([
+  PropTypes.string,
+  PropTypes.arrayOf(
+    PropTypes.oneOfType([PropTypes.shape(SectionShape), PropTypes.string]),
+  ),
+])
+
+SectionShape.children = ValidChildren
+
+MetaMaskTemplateRenderer.propTypes = {
+  sections: ValidChildren,
+}
+
+export default memo(MetaMaskTemplateRenderer)
```

Comment:
By my reading of [the docs for `React.memo`](https://reactjs.org/docs/react-api.html#reactmemo), I suspect we'll need to pass a custom comparison function as the second parameter to `memo` to get a significant performance boost:

> By default it will only shallowly compare complex objects in the props object.

If I'm right, I don't want to hold up this PR on that point, and would rather see the call to `memo` removed than cause delays. We can always optimize in a follow-up.

Reply:
Question: The recursive calls are not memoized the way this is written, right?

Reply:
I think this could be an easy win.

#### More background
The current implementations of MetaMaskTemplateRenderer handle memoization themselves, only passing referentially equal arrays to the `sections` prop.

For example, the new confirmation UI page has to generate template values based on the current confirmation. Because calling the template generator function is expensive, it is memoized and therefore the value will be referentially equal as long as the confirmation doesn't change.

Another scenario we might see is where the template is just a `const` and doesn't need to change. In which case, we don't need the isEqual check on the `memo` call.

Another case is we might be receiving a template payload from another provider, be it our own background or a third party in the future. In those cases, we'll be deserializing the template in some form or fashion and will want to handle memory allocation specifically. 

#### lodash.isEqual
The very first thing this function does is check referential equality, and then it ejects early if they are equal. So even though the usages of the component might be passing referentially equal objects, having isEqual as the comparator function will give easy optimization for future usages that don't care so much about the values they are passing in. It also would mean that if the memoized values are updated but the resulting object is the same, this would also prevent a re-render.

#### Recursive calls
They should bypass the memo method as written, which was done purposefully to allow the root node of the tree to be the sole manager of re-rendering.